### PR TITLE
Add create app to CLI and fix delete

### LINKED ISSUE #515: Epinio app create CLI subcommand
Using the new CLI command we can create an app, without triggering staging to deploy a workload in the cluster.

## Acceptance Criteria

* We use the new application create endpoint from the CLI

### PR DESCRIPTION
**Pull request by manno**
With the new apps create command, it was also necessary to fix apps delete, so it can delete apps that don't have a workload.

fixes #515

### CODE REVIEW

**Comment by manno** [comment]
CI is flaky:

```
[7]       🕞  [5fd97cf3c5fd3120-stage-b6vgw-pod-78jtx] step-create Saving epinio-registry.192.168.64.3.omg.howdoi.website/apps/apps-932472169-53bbef6aade4515f83cfabf17a27e08385a868fa... 
[7]       🕞  [5fd97cf3c5fd3120-stage-b6vgw-pod-78jtx] step-create *** Images (sha256:056d7052ff8b4e5f9ab26e46758386e955a57d71730d1ff45dd7bad68432864b): 
[7]       🕞  [5fd97cf3c5fd3120-stage-b6vgw-pod-78jtx] step-create       epinio-registry.192.168.64.3.omg.howdoi.website/apps/apps-932472169-53bbef6aade4515f83cfabf17a27e08385a868fa - GET https://epinio-registry.192.168.64.3.omg.howdoi.website/v2/: unexpected status code 404 Not Found: 404 page not found 
[7]       🕞  [5fd97cf3c5fd3120-stage-b6vgw-pod-78jtx] step-create  
[7]       🕞  [5fd97cf3c5fd3120-stage-b6vgw-pod-78jtx] step-create ERROR: failed to export: failed to write image to the following tags: [epinio-registry.192.168.64.3.omg.howdoi.website/apps/apps-932472169-53bbef6aade4515f83cfabf17a27e08385a868fa: GET https://epinio-registry.192.168.64.3.omg.howdoi.website/v2/: unexpected status code 404 Not Found: 404 page not found 
[7]       🕞  [5fd97cf3c5fd3120-stage-b6vgw-pod-78jtx] step-create ] 
[7]       🕞  [5fd97cf3c5fd3120-stage-b6vgw-pod-78jtx] step-results 2021/06/28 15:21:32 Skipping step because a previous step failed 
[7]       error pushing app to server: waiting for staging failed: Tasks Completed: 2 (Failed: 1, Cancelled 0), Skipped: 2
[7]       
[7]       Unexpected error:
[7]           <*exec.ExitError | 0xc00007c000>: {
[7]               ProcessState: {
[7]                   pid: 12079,
[7]                   status: 65280,
[7]                   rusage: {
[7]                       Utime: {Sec: 0, Usec: 288747},
[7]                       Stime: {Sec: 0, Usec: 28401},
[7]                       Maxrss: 43724,
[7]                       Ixrss: 0,
[7]                       Idrss: 0,
[7]                       Isrss: 0,
[7]                       Minflt: 2421,
[7]                       Majflt: 0,
[7]                       Nswap: 0,
[7]                       Inblock: 0,
[7]                       Oublock: 136,
[7]                       Msgsnd: 0,
[7]                       Msgrcv: 0,
[7]                       Nsignals: 0,
[7]                       Nvcsw: 7451,
[7]                       Nivcsw: 76,
[7]                   },
[7]               },
[7]               Stderr: nil,
[7]           }
[7]           exit status 255
[7]       occurred
[7] 
[7]       /home/carrier/action-runner/_work/epinio/epinio/acceptance/utilities_test.go:101
```


```
[2]     🚢 Create Service
[2]     Name: service-560912905
[2]     Organization: orgs-560911555
[2]     Class: mariadb
[2]     Plan: 10-3-22
[2]     | PARAMETER | VALUE |
[2]     |-----------|-------|
[2]     
[2]     🚢 Provisioning................................................................................................................................................................................................................................................................................................................error creating service: Internal Server Error: {"errors":[{"status":500,"title":"timed out waiting for the condition","details":""}]}
[2]     
[2]     
[2]     Unexpected error:
[2]         <*exec.ExitError | 0xc000806000>: {
[2]             ProcessState: {
[2]                 pid: 24959,
[2]                 status: 65280,
[2]                 rusage: {
[2]                     Utime: {Sec: 0, Usec: 93496},
[2]                     Stime: {Sec: 0, Usec: 27702},
[2]                     Maxrss: 47296,
[2]                     Ixrss: 0,
[2]                     Idrss: 0,
[2]                     Isrss: 0,
[2]                     Minflt: 1553,
[2]                     Majflt: 0,
[2]                     Nswap: 0,
[2]                     Inblock: 0,
[2]                     Oublock: 0,
[2]                     Msgsnd: 0,
[2]                     Msgrcv: 0,
[2]                     Nsignals: 0,
[2]                     Nvcsw: 3788,
[2]                     Nivcsw: 22,
[2]                 },
[2]             },
[2]             Stderr: nil,
[2]         }
[2]         exit status 255
[2]     occurred
[2] 
[2]     /home/carrier/action-runner/_work/epinio/epinio/acceptance/catalog_services_test.go:47
```

File: internal/application/application.go [hunk 1]
```diff
@@ -73,7 +107,7 @@ func Lookup(ctx context.Context, cluster *kubernetes.Cluster, org, lookupApp str
 	return nil, nil
 }
 
-// List returns an list of all available applications (in the org)
+// List returns an list of all available workloads (in the org)
```
Comment (andreas-kupries) [inline internal/application/application.go:110]:
```suggestion
// List returns a list of all available workloads (in the org)
```
Absolutely a nit.

File: internal/cli/clients/apps_create.go [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+import (
+	"encoding/json"
+
+	api "github.com/epinio/epinio/internal/api/v1"
+	"github.com/epinio/epinio/internal/api/v1/models"
+)
+
+// AppCreate updates the credentials stored in the config from the
+// currently targeted kube cluster
```
Comment (andreas-kupries) [inline internal/cli/clients/apps_create.go:11]: `AppCreate` has nothing to do with credentials and config.

**Review by andreas-kupries — CHANGES_REQUESTED** [review]
Some nits in comment so far.

**Comment by andreas-kupries** [comment]
Ok, as you fixed `app delete` I suspect that fixing `app list` will be in scope also.

```
work@tagetarl:~/SUSE/dev/epinio> ep app list 

🚢 Listing applications
Organization: workspace

✔️  Epinio Applications:
| NAME | STATUS | ROUTES | SERVICES |
|------|--------|--------|----------|
```

This is after `ep app create hello` was run successfully.

**Comment by andreas-kupries** [comment]
I wonder if that cross over into #536 territory

**Comment by andreas-kupries** [comment]
And with the empty app present push fails also. Considers the apps to exist.

```
work@tagetarl:~/SUSE/dev/epinio> ep app push hello assets/sample-app

🚢 About to push an application with given name and sources into the specified organization
Name: hello
Sources: assets/sample-app
Organization: workspace

⚠️  Hit Enter to continue or Ctrl+C to abort (deployment will continue automatically in 5 seconds)


Create the application resource ...
error pushing app to server: Internal Server Error: {"errors":[{"status":500,"title":"applications.app.k8s.io \"hello\" already exists","details":""}]}
```

After deleting the lone resource it is ok to push again.

Side note, back to `app list` ... Related to that command is the match function for argument completion.

**Comment by manno** [comment]
> Ok, as you fixed `app delete` I suspect that fixing `app list` will be in scope also.

This is actually a separate story: #510 
I had to fix delete to make the test green. Then I had so much context on deletion that I also tried to fix the "deleting incomplete deployments" problem. 

> And with the empty app present push fails also. Considers the app to exist at a point where it does not consider updating as solution

Good catch. That should have been in the acceptance criteria. This feature doesn't make sense without being able to push.

File: internal/cli/apps.go [hunk 3]
```diff
@@ -62,6 +63,43 @@ var CmdAppList = &cobra.Command{
...
+
+		err = client.AppCreate(args[0])
+		if err != nil {
+			return errors.Wrap(err, "error creating app")
+		}
+
+		return nil
+	},
+	SilenceErrors: true,
+	SilenceUsage:  true,
```
Comment (andreas-kupries) [inline internal/cli/apps.go:86]: This should be adapted to the changes in the other commands, i.e. `Silence*: false` by default, and `SilenceUsage` dynamically activated in the command function. See PR #539 . This is something which can be done after this is merged. I.e. a nit, and I do not wish to cause more delay.

**Review by andreas-kupries — APPROVED** [review]
Found a nit around the cli help for this command. Approved anyway, can be fixed at leisure, should not delay merge of the functionality itself.

**Pull request merged**